File: sakila-payment-service/src/main/java/com/example/sakila/module/customer/repository/CustomerRepositoryHibernateImpl.java
package com.example.sakila.module.customer.repository;

import com.example.sakila.module.customer.Customer;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.TypedQuery;
import javax.transaction.Transactional;
import java.util.List;

@Repository
public class CustomerRepositoryHibernateImpl implements CustomerRepository {

  @PersistenceContext
  private EntityManager entityManager;

  @Override
  public Customer getCustomerById(Long id) {
    return entityManager.find(Customer.class, id);
  }

  @Override
  public List<Customer> getCustomersByStoreId(Long id) {
    TypedQuery<Customer> query = createQuery("SELECT c FROM Customer c WHERE c.store_id = :storeId");
    query.setParameter("storeId", id);
    return query.getResultList();
  }

  @Override
  public List<Customer> searchCustomersByFirstName(String expression) {
    TypedQuery<Customer> query = createQuery(
        "SELECT c FROM Customer c WHERE LOWER(c.firstName) LIKE LOWER(:expression)"
    );
    query.setParameter("expression", '%' + expression + '%');
    return query.getResultList();
  }

  @Override
  public List<Customer> searchCustomersByLastName(String expression) {
    TypedQuery<Customer> query = createQuery(
        "SELECT c FROM Customer c WHERE LOWER(c.lastName) LIKE LOWER(:expression)"
    );
    query.setParameter("expression", '%' + expression + '%');
    return query.getResultList();
  }

  @Override
  @Transactional
  public Customer insertCustomer(Customer customer) {
    entityManager.persist(customer);
    entityManager.flush();
    entityManager.refresh(customer);
    return customer;
  }

  @Override
  @Transactional
  public Customer updateCustomer(Customer customer) {
    entityManager.merge(customer);
    entityManager.flush();
    entityManager.refresh(customer);
    return customer;
  }

  @Override
  @Transactional
  public void deleteCustomer(Customer customer) {
    entityManager.remove(customer);
  }

  private TypedQuery<Customer> createQuery(String sql) {
    return entityManager.createQuery(sql, Customer.class);
  }
}

File: sakila-payment-service/src/main/java/com/example/sakila/module/customer/repository/CustomerRepositoryHibernateImpl.kt
package com.example.sakila.module.customer.repository

import com.example.sakila.module.customer.Customer
import org.springframework.stereotype.Repository
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext
import javax.persistence.TypedQuery
import javax.transaction.Transactional

@Repository
class CustomerRepositoryHibernateImpl : CustomerRepository {

  @PersistenceContext
  private lateinit var entityManager: EntityManager

  override fun getCustomerById(id: Long): Customer? = entityManager.find(Customer::class.java, id)

  override fun getCustomersByStoreId(id: Long): List<Customer>? {
    val query = createQuery("SELECT c FROM Customer c WHERE c.storeId = :storeId")
    query.setParameter("storeId", id)
    return query.resultList
  }

  override fun searchCustomersByFirstName(firstName: String): List<Customer>? {
    val query = createQuery("SELECT c FROM Customer c WHERE LOWER(c.firstName) LIKE LOWER(:expression)")
    query.setParameter("expression", "%$firstName%")
    return query.resultList
  }

  override fun searchCustomersByLastName(lastName: String): List<Customer>? {
    val query = createQuery("SELECT c FROM Customer c WHERE LOWER(c.lastName) LIKE LOWER(:expression)")
    query.setParameter("expression", "%$lastName%")
    return query.resultList
  }

  @Transactional
  override fun insertCustomer(customer: Customer): Customer {
    entityManager.persist(customer)
    entityManager.flush()
    entityManager.refresh(customer)
    return customer
  }

  @Transactional
  override fun updateCustomer(customer: Customer): Customer {
    entityManager.merge(customer)
    entityManager.flush()
    entityManager.refresh(customer)
    return customer
  }

  @Transactional
  override fun deleteCustomer(customer: Customer) = entityManager.remove(customer)

  private fun createQuery(sql: String): TypedQuery<Customer> = entityManager.createQuery(sql, Customer::class.java)

}
